### PR TITLE
metamorphic: increase probability of valid SeekPrefixGE calls

### PR DESCRIPTION
**Pull request by bananabrick**
This pr increases the probablity of SeekPrefixGE calls which are in
range, and also makes it likely that the key passed to the
SeekPrefixGE function is readable by the iterator.

Currently, we only increase the probablity of valid SeekPrefixGE
calls for iterators created on the DB.

### CODE REVIEW

File: internal/metamorphic/meta_test.go [hunk 1]
```diff
@@ -323,7 +323,7 @@ func TestMeta(t *testing.T) {
 
 		cmd := exec.Command(os.Args[0], args...)
 		out, err := cmd.CombinedOutput()
-		if err != nil {
+		if err != nil || err == nil {
```
Comment (bananabrick) [inline internal/metamorphic/meta_test.go:326]: Debug code. Will remove once the logic here is considered acceptable.

**Comment by bananabrick** [comment]
Increase in valid calls after this change:

```
These are valid:
arjunnair@crlMBP-C02FR0UTMD6TMjA4 metamorphic % grep 'SeekPrefixGE.*true' test | wc -l
    7680

These return false:
arjunnair@crlMBP-C02FR0UTMD6TMjA4 metamorphic % grep 'SeekPrefixGE.*false' test | wc -l
   10224

This is the valid + invalid:
arjunnair@crlMBP-C02FR0UTMD6TMjA4 metamorphic % grep 'SeekPrefixGE.*//' test | wc -l   
   17904
arjunnair@crlMBP-C02FR0UTMD6TMjA4 metamorphic % 
```

**Review by bananabrick — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1922)* status: 0 of 3 files reviewed, 2 unresolved discussions (waiting on @jbowens)

___
*[`internal/metamorphic/generator.go` line 37 at r1](https://reviewable.io/reviews/cockroachdb/pebble/1922#-NA5rlzlC9GGMjS_pbwF:-NA5rlzlC9GGMjS_pbwG:b52vfik) ([raw file](https://github.com/cockroachdb/pebble/blob/2efd34c0724486524bccc74296662c111f673635/internal/metamorphic/generator.go#L37)):*
> ```Go
> }
> 
> // metaTimestamp is used to provide a ordering over certain operations like
> ```

I had to introduce this concept to determine whether an iterator could view a key. We keep track of the timestamp at which an iterator is created, and the timestamp at which certain updates to keys happen. We need to determine whether an iterator can read a key 

Another approach is to iterate over valid keys on iter creation, and use save some of those keys for SeekPrefixGE calls. However, a call to SetBounds can make some of those keys invalid.

**Comment by bananabrick** [comment]
Open to other approaches too.

**Comment by bananabrick** [comment]
Another method would be to generate keys which can't be deleted with some probability on key generation. Then, we can use those for SeekPrefixGE calls.

**Review by jbowens — COMMENTED** [review]
> Reviewed 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1922)* status: 1 of 3 files reviewed, 8 unresolved discussions (waiting on @bananabrick and @jbowens)

___
*[`internal/metamorphic/generator.go` line 40 at r2](https://reviewable.io/reviews/cockroachdb/pebble/1922#-NAAVPc8ERNNPp4KZ0lY:-NAAVPc8ERNNPp4KZ0lZ:brjiw0c) ([raw file](https://github.com/cockroachdb/pebble/blob/4d4e25cc1a36d6bd6c2c8ee0ea2466651823cf50/internal/metamorphic/generator.go#L40)):*
> ```Go
> // iter creation, updates to keys. Keeping track of the timestamp allows us to
> // make determinations such as whether a key will be visible to an iterator.
> var metaTimestamp = 0
> ```

can we lift this off the global scope and onto the `generator`? wdyt of adding a type to disambiguate, eg `type metaTimestamp int`.
___
*[`internal/metamorphic/generator.go` line 854 at r2](https://reviewable.io/reviews/cockroachdb/pebble/1922#-NAAXpT01XOzITPjZzIC:-NAAXpT01XOzITPjZzID:b-2ttmcw) ([raw file](https://github.com/cockroachdb/pebble/blob/4d4e25cc1a36d6bd6c2c8ee0ea2466651823cf50/internal/metamorphic/generator.go#L854)):*
> ```Go
> 	upper := g.itersLastOpts[iterID].upper
> 	iterCreationTimestamp := g.iterCreationTimestamp[iterID]
> 	possibleKeys := make([][]byte, 0, 100)
> ```

nit: move alloc into the `if` statement below, and rely on `key != nil` down below to determine whether we call `g.randKeyToRead`.
___
*[`internal/metamorphic/generator.go` line 859 at r2](https://reviewable.io/reviews/cockroachdb/pebble/1922#-NAAY9m651CmdvMANWKp:-NAAY9m651CmdvMANWKq:bu4j4io) ([raw file](https://github.com/cockroachdb/pebble/blob/4d4e25cc1a36d6bd6c2c8ee0ea2466651823cf50/internal/metamorphic/generator.go#L859)):*
> ```Go
> 	// SeekPrefixGE key is within the iter bounds, and that the iter can read
> 	// the key.
> 	if g.iterReaderID[iterID] == dbObjID {
> ```

can we make this case also probabilistic so that we sometimes use a completely random key? eg, `g.iterReaderID[iterID] == dbObjID && g.rng.Intn(5) >= 1`
___
*[`internal/metamorphic/generator.go` line 885 at r2](https://reviewable.io/reviews/cockroachdb/pebble/1922#-NAAZYI490Oz0EbRNVYw:-NAAZYI490Oz0EbRNVYx:b-1mmw53) ([raw file](https://github.com/cockroachdb/pebble/blob/4d4e25cc1a36d6bd6c2c8ee0ea2466651823cf50/internal/metamorphic/generator.go#L885)):*
> ```Go
> 		key = []byte(possibleKeys[g.rng.Int31n(int32(len(possibleKeys)))])
> 	} else {
> 		key = g.randKeyToRead(0) // 0% new keys
> ```

can we separate the bounds enforcement from the attempt to increase probability from liveness? eg, add a `g.randKeyToReadWithinBounds` func?
___
*[`internal/metamorphic/key_manager.go` line 60 at r2](https://reviewable.io/reviews/cockroachdb/pebble/1922#-NAAWyXa4oKewGwhS7Sz:-NAAWyXa4oKewGwhS7T-:bqjwd0m) ([raw file](https://github.com/cockroachdb/pebble/blob/4d4e25cc1a36d6bd6c2c8ee0ea2466651823cf50/internal/metamorphic/key_manager.go#L60)):*
> ```Go
> 
> 	// updateOps should always be ordered by non-decreasing metaTimestamp.
> 	// updateOps will not be updated if they key is range deleted. Therefore, it
> ```

s/they key/the key/
___
*[`internal/metamorphic/key_manager.go` line 103 at r2](https://reviewable.io/reviews/cockroachdb/pebble/1922#-NAAW8571Pu3c2YR8rrs:-NAAW8571Pu3c2YR8rrt:bg3npev) ([raw file](https://github.com/cockroachdb/pebble/blob/4d4e25cc1a36d6bd6c2c8ee0ea2466651823cf50/internal/metamorphic/key_manager.go#L103)):*
> ```Go
> 	}
> 
> 	// Determine if they key is visible or not after the ketMetas are merged.
> ```

s/they key/the key/

s/ketMetas/keyMetas/

**Review by bananabrick — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1922)* status: 0 of 3 files reviewed, 8 unresolved discussions (waiting on @jbowens)

___
*[`internal/metamorphic/generator.go` line 40 at r2](https://reviewable.io/reviews/cockroachdb/pebble/1922#-NAAVPc8ERNNPp4KZ0lY:-NAB0W_i0YGuHc3IQwRc:boqrv0x) ([raw file](https://github.com/cockroachdb/pebble/blob/4d4e25cc1a36d6bd6c2c8ee0ea2466651823cf50/internal/metamorphic/generator.go#L40)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

can we lift this off the global scope and onto the `generator`? wdyt of adding a type to disambiguate, eg `type metaTimestamp int`.
</blockquote></details>

The keymanager uses and generates metatimestamps, but it doesn't have access to the generator. I could thread around a function throughout the key manager to generate these timestamps. What do you think? 
___
*[`internal/metamorphic/generator.go` line 859 at r2](https://reviewable.io/reviews/cockroachdb/pebble/1922#-NAAY9m651CmdvMANWKp:-NAB5qCz4h0bJ-EU-7r7:b-896fix) ([raw file](https://github.com/cockroachdb/pebble/blob/4d4e25cc1a36d6bd6c2c8ee0ea2466651823cf50/internal/metamorphic/generator.go#L859)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

can we make this case also probabilistic so that we sometimes use a completely random key? eg, `g.iterReaderID[iterID] == dbObjID && g.rng.Intn(5) >= 1`
</blockquote></details>

Done.
___
*[`internal/metamorphic/generator.go` line 885 at r2](https://reviewable.io/reviews/cockroachdb/pebble/1922#-NAAZYI490Oz0EbRNVYw:-NAB6AsJ0qbXnh83vQRs:bnoka3l) ([raw file](https://github.com/cockroachdb/pebble/blob/4d4e25cc1a36d6bd6c2c8ee0ea2466651823cf50/internal/metamorphic/generator.go#L885)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

can we separate the bounds enforcement from the attempt to increase probability from liveness? eg, add a `g.randKeyToReadWithinBounds` func?
</blockquote></details>

In the case where we want to generate keys within bounds which are visible to the iterator, the algorithm is:

1. start with the keys in the db.
2. Check if key is visible to iterator by timestamp
3. Check if key is within bounds. (we can probably swap 2, 3).

If I have a `g.randKeyToReadWithinBounds`, how would I use it in the case where we want to generate keys which the iter can read? Maybe we can have a `g.keysWithinBounds(reader objID) []keyMeta` func instead? I changed the logic a bit. Lmk if you had something else in mind.
___
*[`internal/metamorphic/generator.go` line 824 at r3](https://reviewable.io/reviews/cockroachdb/pebble/1922#-NABNZE-4hvpSs1tGrwR:-NABNZE-4hvpSs1tGrwS:b-zcpg2b) ([raw file](https://github.com/cockroachdb/pebble/blob/9934462109068b3ad5dfa66d2221232678ecb182/internal/metamorphic/generator.go#L824)):*
> ```Go
> 	// keys in a best effort manner, and the logic is better than picking a
> 	// random key.
> 	if g.rng.Intn(10) >= 1 {
> ```

But this a 90% chance instead of 80. There's cases when we don't find keys anyway, so we might not even need that much.
___
*[`internal/metamorphic/key_manager.go` line 60 at r2](https://reviewable.io/reviews/cockroachdb/pebble/1922#-NAAWyXa4oKewGwhS7Sz:-NAB49yNDNEbpoiVhML-:b-896fix) ([raw file](https://github.com/cockroachdb/pebble/blob/4d4e25cc1a36d6bd6c2c8ee0ea2466651823cf50/internal/metamorphic/key_manager.go#L60)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

s/they key/the key/
</blockquote></details>

Done.
___
*[`internal/metamorphic/key_manager.go` line 103 at r2](https://reviewable.io/reviews/cockroachdb/pebble/1922#-NAAW8571Pu3c2YR8rrs:-NAB4Ux_86dt9bsJFnOX:b-896fix) ([raw file](https://github.com/cockroachdb/pebble/blob/4d4e25cc1a36d6bd6c2c8ee0ea2466651823cf50/internal/metamorphic/key_manager.go#L103)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

s/they key/the key/

s/ketMetas/keyMetas/
</blockquote></details>

Done.

**Review by bananabrick — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1922)* status: 0 of 3 files reviewed, 8 unresolved discussions (waiting on @jbowens)

___
*[`internal/metamorphic/generator.go` line 40 at r2](https://reviewable.io/reviews/cockroachdb/pebble/1922#-NAAVPc8ERNNPp4KZ0lY:-NABWGX8A2AJzCmOTJcO:bv9lsct) ([raw file](https://github.com/cockroachdb/pebble/blob/4d4e25cc1a36d6bd6c2c8ee0ea2466651823cf50/internal/metamorphic/generator.go#L40)):*
<details><summary><i>Previously, bananabrick (Arjun Nair) wrote…</i></summary><blockquote>

The keymanager uses and generates metatimestamps, but it doesn't have access to the generator. I could thread around a function throughout the key manager to generate these timestamps. What do you think? 
</blockquote></details>

Aah, I think it would be better to have the metatimestamp managed by the `keyManager`.

**Review by bananabrick — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1922)* status: 0 of 3 files reviewed, 8 unresolved discussions (waiting on @jbowens)

___
*[`internal/metamorphic/generator.go` line 40 at r2](https://reviewable.io/reviews/cockroachdb/pebble/1922#-NAAVPc8ERNNPp4KZ0lY:-NABXBOy6KPYl9u9JDVM:bic6owy) ([raw file](https://github.com/cockroachdb/pebble/blob/4d4e25cc1a36d6bd6c2c8ee0ea2466651823cf50/internal/metamorphic/generator.go#L40)):*
<details><summary><i>Previously, bananabrick (Arjun Nair) wrote…</i></summary><blockquote>

Aah, I think it would be better to have the metatimestamp managed by the `keyManager`.
</blockquote></details>

Although, the timestamp is used in tests which test `keyMeta` where the key manager doesn't even exist, so maybe it's best to keep it global.

**Review by jbowens — COMMENTED** [review]
> Reviewed 2 of 3 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1922)* status: 3 of 4 files reviewed, 5 unresolved discussions (waiting on @bananabrick and @jbowens)

___
*[`internal/metamorphic/generator.go` line 40 at r2](https://reviewable.io/reviews/cockroachdb/pebble/1922#-NAAVPc8ERNNPp4KZ0lY:-NAKL40l05BN7sI2hR47:b-6m9rzr) ([raw file](https://github.com/cockroachdb/pebble/blob/4d4e25cc1a36d6bd6c2c8ee0ea2466651823cf50/internal/metamorphic/generator.go#L40)):*
<details><summary><i>Previously, bananabrick (Arjun Nair) wrote…</i></summary><blockquote>

Although, the timestamp is used in tests which test `keyMeta` where the key manager doesn't even exist, so maybe it's best to keep it global.
</blockquote></details>

Moving onto the `keyManager` sounds good — I think a global is too much of a foot gun.
___
*[`internal/metamorphic/generator.go` line 885 at r2](https://reviewable.io/reviews/cockroachdb/pebble/1922#-NAAZYI490Oz0EbRNVYw:-NAKLq-N0oyhXKDUfWVt:b-qevzwy) ([raw file](https://github.com/cockroachdb/pebble/blob/4d4e25cc1a36d6bd6c2c8ee0ea2466651823cf50/internal/metamorphic/generator.go#L885)):*
<details><summary><i>Previously, bananabrick (Arjun Nair) wrote…</i></summary><blockquote>

In the case where we want to generate keys within bounds which are visible to the iterator, the algorithm is:

1. start with the keys in the db.
2. Check if key is visible to iterator by timestamp
3. Check if key is within bounds. (we can probably swap 2, 3).

If I have a `g.randKeyToReadWithinBounds`, how would I use it in the case where we want to generate keys which the iter can read? Maybe we can have a `g.keysWithinBounds(reader objID) []keyMeta` func instead? I changed the logic a bit. Lmk if you had something else in mind.
</blockquote></details>

`g.keysWithinBounds(reader objID) []keyMeta` sounds good, because it makes picking a random key within bounds trivial.

**Review by bananabrick — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1922)* status: 1 of 4 files reviewed, 6 unresolved discussions (waiting on @bananabrick and @jbowens)

___
*[`internal/metamorphic/generator.go` line 825 at r4](https://reviewable.io/reviews/cockroachdb/pebble/1922#-NAep-S2E2WdxQ8zGdNS:-NAep-S2E2WdxQ8zGdNT:b4i46di) ([raw file](https://github.com/cockroachdb/pebble/blob/01cd8b99dacf2ccfb3e9237b6b76ce5729abe03d/internal/metamorphic/generator.go#L825)):*
> ```Go
> 	// random key.
> 	if g.rng.Intn(10) >= 1 {
> 		possibleKeys := make([][]byte, 0, 100)
> ```

There's ways to get rid of these allocations, and the one in `randKeyToReadWithinBounds`, but seems unnecessarily complicated for metamorphic tests, where we can handle some allocations in the generation phase.

**Review by jbowens — APPROVED** [review]
:lgtm:

> Reviewed 1 of 3 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1922#-:-NAfFDSk3uwrDCwlsd5u:bnfp4nl)* status: 2 of 4 files reviewed, 7 unresolved discussions (waiting on @bananabrick and @jbowens)

___
*[`internal/metamorphic/generator.go` line 772 at r5](https://reviewable.io/reviews/cockroachdb/pebble/1922#-NAfD8J4CRBcUfulBR-H:-NAfD8J4CRBcUfulBR-I:bkokovg) ([raw file](https://github.com/cockroachdb/pebble/blob/f3b522385a81ffdb1302c72f5c94cd4db7d04b47/internal/metamorphic/generator.go#L772)):*
> ```Go
> 	for _, keyMeta := range g.keyManager.byObj[readerID] {
> 		posKey := keyMeta.key
> 		if g.cmp(posKey, lower) < 0 || g.cmp(posKey, upper) > 0 {
> ```

I think this should be `g.cmp(posKey, upper) >= 0` because `upper` is exclusive when used as an iterator bound.
___
*[`internal/metamorphic/generator.go` line 821 at r5](https://reviewable.io/reviews/cockroachdb/pebble/1922#-NAfBTqj9WlRokV9QOtA:-NAfBTqj9WlRokV9QOtB:b-h9brfs) ([raw file](https://github.com/cockroachdb/pebble/blob/f3b522385a81ffdb1302c72f5c94cd4db7d04b47/internal/metamorphic/generator.go#L821)):*
> ```Go
> 
> 	if key == nil {
> 		key = g.randKeyToRead(0) // 0% new keys
> ```

i wish we could use within-bounds keys even if we weren't able to find a `dbObjID` key that didn't appear to be deleted, but it's a little awkward right now. Can you leave a TODO?

**Review by bananabrick — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1922)* status: 2 of 4 files reviewed, 4 unresolved discussions (waiting on @jbowens)

___
*[`internal/metamorphic/generator.go` line 40 at r2](https://reviewable.io/reviews/cockroachdb/pebble/1922#-NAAVPc8ERNNPp4KZ0lY:-NAfh_q843rmkuKnGgDw:b-896fix) ([raw file](https://github.com/cockroachdb/pebble/blob/4d4e25cc1a36d6bd6c2c8ee0ea2466651823cf50/internal/metamorphic/generator.go#L40)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

Moving onto the `keyManager` sounds good — I think a global is too much of a foot gun.
</blockquote></details>

Done.
___
*[`internal/metamorphic/generator.go` line 772 at r5](https://reviewable.io/reviews/cockroachdb/pebble/1922#-NAfD8J4CRBcUfulBR-H:-NAfhaHS4kP3joRGZOPA:b-896fix) ([raw file](https://github.com/cockroachdb/pebble/blob/f3b522385a81ffdb1302c72f5c94cd4db7d04b47/internal/metamorphic/generator.go#L772)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

I think this should be `g.cmp(posKey, upper) >= 0` because `upper` is exclusive when used as an iterator bound.
</blockquote></details>

Done.
___
*[`internal/metamorphic/generator.go` line 821 at r5](https://reviewable.io/reviews/cockroachdb/pebble/1922#-NAfBTqj9WlRokV9QOtA:-NAfhc1WAkoUJ3rzIbhz:b-896fix) ([raw file](https://github.com/cockroachdb/pebble/blob/f3b522385a81ffdb1302c72f5c94cd4db7d04b47/internal/metamorphic/generator.go#L821)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

i wish we could use within-bounds keys even if we weren't able to find a `dbObjID` key that didn't appear to be deleted, but it's a little awkward right now. Can you leave a TODO?
</blockquote></details>

Done.

**Pull request merged**